_episodes/l4_04_parameterized_tests.md

```diff
@@ -0,0 +1,353 @@
+
+Moreover, the problem gets worse when one of the test fails. In order to understand what happens during a test failure when using a for loop, let us intentionally change the expected output value to an incorrect value. In file, [2_Test_using_for_loop.cpp](../code/Chapter4/2_Test_using_for_loop.cpp), you can make the following change.
+
+```cpp
+TEST_F(EmployeeTestFixture, NetBonusIsCorrectForDifferentYears) {
+    auto input = vector<int>{5, 15};
+    auto expected_output = vector<int>{2000, 7000};
+    for (int i = 0; i < input.size(); i++) {
+        employee.setNumberYearsEmployed(input[i]);
+        EXPECT_EQ(employee.getNetBonus(), expected_output[i]);
+    }
+}
+```
+
+On running the code with this change, we get the following output.
+
+```bash
+[==========] Running 1 test from 1 test suite.
+[----------] Global test environment set-up.
+[----------] 1 test from EmployeeTestFixture
+[ RUN      ] EmployeeTestFixture.NetBonusIsCorrectForDifferentYears
+2_Test_using_for_loop.cpp:21: Failure
+Expected equality of these values:
+  employee.getNetBonus()
+    Which is: 3000
+  expected_output[i]
+    Which is: 7000
+[  FAILED  ] EmployeeTestFixture.NetBonusIsCorrectForDifferentYears (0 ms)
+[----------] 1 test from EmployeeTestFixture (0 ms total)
+
+[----------] Global test environment tear-down
+[==========] 1 test from 1 test suite ran. (0 ms total)
+[  PASSED  ] 0 tests.
+[  FAILED  ] 1 test, listed below:
+[  FAILED  ] EmployeeTestFixture.NetBonusIsCorrectForDifferentYears
+
+ 1 FAILED TEST
+```
+
+From the output, we can clearly see that it results in complete failure of the test even though one of the conditions (or test) was right. Moreover, the output does not help much to figure out which test has exactly failed.
+
+The solution for above mentioned issues is to make use of parameterised tests and the next section describes that.
+
+## 2. Parameterised tests in GoogleTest
+
+In Google Test, parameterised tests are implemented using the `TEST_P` macro, where "P" stands for parameterised. We define a test class and then specify multiple sets of input data using the `INSTANTIATE_TEST_CASE_P` macro. Each set of input data represents a different instance of the test, and the test framework runs the test case for each instance.
+
+A parameterized tests in GoogleTest requires the following components in general.
+
+1. **A parameterised test class**: Similar to the process of a test fixture, we need to create a class derived from `testing::TestWithParam<T>` where `T` could be any valid C++ type.
+
+```cpp
+class YourTestParameterisedClass : public::TestWithParam<T> {
+    public:
+        ClassUnderTest publicInstance;
+};
+```
```

Comment:
So, a question form my side: can you only run parametrised tests in classes and not in top level functions?

Reply:
@dalonsoa Thanks for this question. We can run parameterized test in any function and the procedure is similar and straightforward.

Here is the code to illustrate that
```cpp
#include <gtest/gtest.h>
#include <vector>

// Define your normal function
int Sum(int a, int b) {
    return a + b;
}

// Define a test fixture class
class ParameterizedTest : public testing::TestWithParam<std::pair<int, int>> {
};

// Define the test case with the parameterized test
TEST_P(ParameterizedTest, TestSum) {
    // Get the parameter values
    int a = GetParam().first;
    int b = GetParam().second;

    // Call your normal function
    int result = Sum(a, b);

    // Perform assertions
    ASSERT_EQ(a + b, result);
}

// Define the test data
INSTANTIATE_TEST_SUITE_P(Default, ParameterizedTest, testing::Values(
    std::make_pair(1, 1),
    std::make_pair(2, 3),
    std::make_pair(-5, 10)
));

int main(int argc, char** argv) {
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
```